File: app/src/main/java/com/agrogames/islandsofwar/graphics/impl/drawtexture/TextureDrawer.java
package com.agrogames.islandsofwar.graphics.impl.drawtexture;

import android.graphics.Bitmap;
import android.os.Build;
import android.util.Pair;

import androidx.annotation.RequiresApi;

import com.agrogames.islandsofwar.graphics.impl.bitmap.BitmapDescriptor;
import com.agrogames.islandsofwar.graphics.impl.bitmap.BitmapProvider;
import com.agrogames.islandsofwar.graphics.impl.gl.Texture;

import java.util.ArrayList;
import java.util.List;

public class TextureDrawer implements com.agrogames.islandsofwar.graphics.abs.TextureDrawer {
    private final List<Texture> textures = new ArrayList<>();
    private final BitmapProvider bitmapProvider;
    private float currentX;
    private float currentY;
    private float currentZoom = 1;

    public TextureDrawer(BitmapProvider bitmapProvider) {
        this.bitmapProvider = bitmapProvider;
    }

    @Override
    public void drawTexture(float x, float y, String bitmap, float width, float height, float rotation) {
        textures.add(new Texture(x * currentZoom + currentX, y * currentZoom + currentY, bitmapProvider.load(bitmap).id, width * currentZoom, height * currentZoom, rotation));
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    public Pair<Float, Float> drawTexture(float x, float y, String bitmap, float rotation) {
        BitmapDescriptor b = bitmapProvider.load(bitmap);
        float width = b.width / 50f;
        float height = b.height / 50f;
        textures.add(new Texture(x * currentZoom + currentX, y * currentZoom + currentY, b.id, width * currentZoom, height * currentZoom, rotation));
        return new Pair<>(width, height);
    }

    @Override
    public Pair<Float, Float> getSize(String bitmap) {
        BitmapDescriptor b = bitmapProvider.load(bitmap);
        float width = b.width / 50f;
        float height = b.height / 50f;
        return new Pair<>(width, height);
    }

    @Override
    public void translate(float x, float y) {
        currentX += x;
        currentY += y;
    }

    @Override
    public void scale(float zoom) {
        currentZoom *= zoom;
    }

    public Texture[] GetTextures(){
        Texture[] result = textures.toArray(new Texture[0]);
        textures.clear();
        return result;
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/graphics/drawtexture/TextureDrawer.kt
package com.agrogames.islandsofwar.graphics.drawtexture

import android.util.Pair
import com.agrogames.islandsofwar.graphics.bitmap.BitmapProvider
import com.agrogames.islandsofwar.graphics.gl.Texture

class TextureDrawer {
    private val textures: MutableList<Texture> = ArrayList()
    private var currentX = 0f
    private var currentY = 0f
    private var currentZoom = 1f
    fun drawTexture(x: Float, y: Float, bitmap: String, width: Float, height: Float, rotation: Float) {
        textures.add(Texture(x * currentZoom + currentX, y * currentZoom + currentY, BitmapProvider.load(bitmap).id, width * currentZoom, height * currentZoom, rotation))
    }

    fun drawTexture(x: Float, y: Float, bitmap: String, rotation: Float): Pair<Float, Float> {
        val b = BitmapProvider.load(bitmap)
        val width = b.width / 50f
        val height = b.height / 50f
        textures.add(Texture(x * currentZoom + currentX, y * currentZoom + currentY, b.id, width * currentZoom, height * currentZoom, rotation))
        return Pair(width, height)
    }

    fun getSize(bitmap: String): Pair<Float, Float> {
        val b = BitmapProvider.load(bitmap)
        val width = b.width / 50f
        val height = b.height / 50f
        return Pair(width, height)
    }

    fun translate(x: Float, y: Float) {
        currentX += x
        currentY += y
    }

    fun scale(currentZoom: Float) {
        this.currentZoom *= currentZoom
    }

    fun GetTextures(): Array<Texture> {
        val result = textures.toTypedArray()
        textures.clear()
        return result
    }
}